pennylane/tape/interfaces/autograd.py

```diff
@@ -126,7 +119,15 @@ def _execute(self, params, device):
         params = [p.item() if p.shape == tuple() else p for p in params]
         params = autograd.builtins.tuple(params)
 
+        # unwrap constant parameters
+        all_params_unwrapped = [
+            p.numpy() if isinstance(p, np.tensor) else p for p in self._all_parameter_values
+        ]
+
+        # evaluate the tape
+        self.set_parameters(all_params_unwrapped, trainable_only=False)
         res = self.execute_device(params, device=device)
+        self.set_parameters(self._all_parameter_values, trainable_only=False)

```

Comment:
Just for my understanding, what are the two new lines for?

Reply:
Ah! These two additional lines are the main bugfix.

Previously, `res = self.execute_device(params, device=device)` would execute the device with the (new) trainable parameters. These would be unwrapped, so that a `ndarray` would be passed to the device rather than a `np.tensor`.

However, this did not consider the edge case where _non_-trainable parameters needed to be unwrapped. This explains why `requires_grad=True` worked, whereas `requires_grad=False` didn't!

Here, we are unwrapping _all_ parameter on the tape, executing it, and then reverting to the original parameters (so there are no side-effects for the user)

Reply:
I think this can be made more efficient in the long-term, by making the two new `set_parameter` calls the _only_ place where parameter setting happens, and remove the analogous logic (for only trainable parameters) from `execute_device`.